### PR TITLE
Allow propagation of errors from Subscriptions channels into Request.…

### PR DESCRIPTION
**Pull request by billettc**
Please see open issue: 
https://github.com/graph-gophers/graphql-go/issues/314

### CODE REVIEW

File: graphql.go [hunk 1]
```diff
@@ -69,6 +69,7 @@ type Schema struct {
 	logger                log.Logger
 	useStringDescriptions bool
 	disableIntrospection  bool
+	prefixRootFunctions   bool
```
Comment (pavelnikolov) [inline graphql.go]: Why do we need this?

**Comment by pavelnikolov** [comment]
Please, resolve the conflicts.

File: graphql.go [hunk 2]
```diff
@@ -98,6 +99,13 @@ func MaxDepth(n int) SchemaOpt {
 	}
 }
 
+// Add the Query, Subscription and Mutation prefixes to the root resolver function when doing reflection from schema to Go code.
+func PrefixRootFunctions() SchemaOpt {
+	return func(s *Schema) {
+		s.prefixRootFunctions = true
+	}
+}
```
Comment (pavelnikolov) [inline graphql.go]: Is this change needed in order to propagate subscription errors?

**Comment by pavelnikolov** [comment]
bump

File: errors/errors.go [hunk 3]
```diff
@@ -4,6 +4,10 @@ import (
 	"fmt"
 )
 
+type SubscriptionError interface {
```
Comment (pavelnikolov) [inline errors/errors.go]: Why do we need this interface?

File: graphql.go (hunk 2 above)
Comment (pavelnikolov) [reply graphql.go]: Why do we need this?

File: internal/exec/packer/packer.go [hunk 4]
```diff
@@ -225,8 +225,9 @@ func (p *StructPacker) Pack(value interface{}) (reflect.Value, error) {
 	for _, f := range p.fields {
 		if value, ok := values[f.field.Name.Name]; ok {
 			packed, err := f.fieldPacker.Pack(value)
+
```
Comment (pavelnikolov) [inline internal/exec/packer/packer.go]: Unnecessary empty line, Please, remove it.

File: internal/exec/resolvable/resolvable.go [hunk 5]
```diff
@@ -229,17 +232,23 @@ func (b *execBuilder) makeObjectExec(typeName string, fields schema.FieldList, p
 	Fields := make(map[string]*Field)
 	rt := unwrapPtr(resolverType)
 	for _, f := range fields {
+
```
Comment (pavelnikolov) [inline internal/exec/resolvable/resolvable.go]: No need for this empty line. Please, remove.

File: internal/exec/subscribe.go [hunk 6]
```diff
@@ -23,17 +23,22 @@ type Response struct {
...
 
+		if len(r.Errs) > 0 {
+			errs = r.Errs
+			return
+		}
+
 		// TODO: move this check into validation.Validate
 		if len(fields) != 1 {
-			err = errors.Errorf("%s", "can subscribe to at most one subscription at a time")
+			errs = []*errors.QueryError{errors.Errorf("%s", "can subscribe to at most one subscription at a time")}
```
Comment (pavelnikolov) [inline internal/exec/subscribe.go]: Under what circumstances can we have more than one error?

**Comment by abourget** [comment]
wow I'm sorry I didn't tend to this earlier.. I'll try to review tomorrow.

Unfortunately, there are two changes combined in this PR.. one is for prefixes on root elements (adds Query and Subscription and Mutation, to the top-level function names), the other is the subscriptions propagation of errors.

I'll get back to you, and perhaps split both.

File: graphql.go (hunk 1 above)
Comment (pavelnikolov) [reply graphql.go]: Yes, please. I would prefer one logical change per PR. Please, remove the prefixes of root functions from this PR.

**Comment by pavelnikolov** [comment]
I want to review the changes separately. Please, send the error propagation changes only.

**Comment by pavelnikolov** [comment]
ping

**Comment by abourget** [comment]
hey @pavelnikolov .. is there a way I can reach out to you privately? Perhaps on Twitter, I connected (@bourgetalexndre).

**Comment by pavelnikolov** [comment]
Yes, I'm usually active in gophers.slack.com - same nickname.

File: internal/exec/subscribe.go [hunk 7]
```diff
@@ -139,8 +139,12 @@ func (r *Request) Subscribe(ctx context.Context, s *resolvable.Schema, op *query
 				}
 				var out bytes.Buffer
 				func() {
-					// TODO: configurable timeout
-					subCtx, cancel := context.WithTimeout(ctx, time.Second)
+					timeout := r.SubscribeResolverTimeout
+					if timeout == 0 {
+						timeout = time.Second
+					}
```
Comment (pavelnikolov) [inline internal/exec/subscribe.go]: I like this one and I'd be happy to merge it. Not sure how I've missed the hard-coded `time.Second`. I'll cherry pick it in a separate PR, though.

**Comment by rawagner** [comment]
@billettc @abourget any chance to move forward with this PR ? I'd really like to have a way to propagate the errors. Can you please split the changes to get them merged soon ?

File: internal/exec/subscribe.go (hunk 7 above)
Comment (maoueh) [reply internal/exec/subscribe.go:145]: Extracted into it's own PR #418

**Comment by maoueh** [comment]
@rawagner @pavelnikolov Extracted into #419

**Comment by maoueh** [comment]
Almost everything has been split into separated laser-focus PRs. I'll do a last pass tomorrow and close this PR since it has been split into smaller chunks.

**Comment by maoueh** [comment]
Ok I just checked and the other issue that we had a fix in this branch has been fixed already by c80e62556cd9cf9e30b9eb4, so there is nothing worth left in this integration branch, I'm going to close this PR, split in multiple smaller ones:
- #418 
- #419 
- #420 
- #421

**Comment by billettc** [comment]
split into multiple pull requests
